File: matlab/functions/project_simplex.m
%% Projection a D-dimensional vector into the probability simplex

function X = project_simplex(Y)
[N,D] = size(Y);
X = sort(Y,2, 'descend');
Xtmp = (cumsum(X,2)-1)*diag(sparse(1./(1:D)));
X = max(bsxfun(@minus,Y,Xtmp(sub2ind([N,D],(1:N)',sum(X>Xtmp,2)))),0);

File: python/functions/project_simplex.py
# Projection of a D-dimension vector into the probability simplex
import numpy as np

def project_simplex(Y):
    n = Y.size
    X = np.array(sorted(Y, reverse = True))
    Xtemp = (np.cumsum(X) - 1) * (1/np.arange(1, n+1))
    e = Xtemp[np.sum(X > Xtemp) - 1]
    f = Y - e
    X = np.maximum(f, 0)
    return X
        
        
        